LongestPole/Main.lean

```diff
@@ -0,0 +1,208 @@
+    else
+      let parents := (graph.find? n).getD #[]
+      let m := parents.foldr (init := m) fun n' m => go n' m
+      let t := (parents.map fun n' => (m.find? n').getD 0).foldr max 0
+      m.insert n ((instructions.find? n).getD 0 + t)
+
+def slowestParents (cumulative : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Name :=
+  graph.fold (init := ∅) fun m n parents =>
+    match parents.toList with
+    | [] => m
+    | h :: t => Id.run do
+      let mut r := h
+      for n' in t do
+        if (cumulative.find? n').getD 0 > (cumulative.find? r).getD 0 then
+          r := n'
+      return m.insert n r
+
+def totalInstructions (instructions : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Float :=
+  let transitive := graph.transitiveClosure
+  transitive.filterMap fun n s => some <| s.fold (init := (instructions.find? n).getD 0)
+    fun t n' => t + ((instructions.find? n').getD 0)
+
+def String.rightPad (s : String) (n : Nat) (c : Char := ' ') : String :=
+  s ++ String.replicate (n - s.length) c
+
+open IO.FS IO.Process Name in
+/-- Implementation of the longest pole command line program. -/
+def longestPoleCLI (args : Cli.Parsed) : IO UInt32 := do
+  let to := match args.flag? "to" with
+  | some to => to.as! ModuleName
+  | none => `Mathlib -- autodetect the main module from the `lakefile.lean`?
+  searchPathRef.set compile_time_search_path%
+  let _ ← unsafe withImportModules #[{module := to}] {} (trustLevel := 1024) fun env => do
+    let graph := env.importGraph
+    let instructions ← speedCenterInstructions (← runForThisCommit)
+    let cumulative := cumulativeInstructions instructions graph
+    let total := totalInstructions instructions graph
+    let slowest := slowestParents cumulative graph
+    let mut table := #[]
+    let mut n := some to
+    while n != none do
+      let c := cumulative.find! n.get!
+      let t := total.find! n.get!
+      let r := (t / c).toString
+      let r := match r.split (· = '.') with
+      | [a, b] => a ++ "." ++ b.take 2
+      | _ => r
+      table := table.push (n.get!, c/10^6 |>.toUInt64, r)
+      n := slowest.find? n.get!
+    let widest := table.map (·.1.toString.length) |>.toList.maximum?.getD 0
+    IO.println s!"{"file".rightPad widest} | instructions | parallelism"
+    IO.println s!"{"".rightPad widest '-'} | ------------ | -----------"
+    for (name, inst, speedup) in table do
+      IO.println s!"{name.toString.rightPad widest} | {(toString inst).rightPad 12} | x{speedup}"
+  return 0
+
+/-- Setting up command line options and help text for `lake exe pole`. -/
+def pole : Cmd := `[Cli|
+  pole VIA longestPoleCLI; ["0.0.1"]
+  "Calculate the longest pole for building Mathlib (or downstream projects).\n" ++
+  "Use as `lake exe pole` or `lake exe pole --to MyProject.MyFile`.\n\n" ++
+  "Prints a sequence of imports starting at the target.\n" ++
+  "For each file, prints the cumulative instructions (in billions) assuming infinite parallelism,\n" ++
```

Comment:
You probably already want to break after "assuming" and then probably once more after "over" to not be double-broken in a 80 column terminal. Alternatively, you may just not break the sentences at all.

Reply:
Done.